ArduinoCode/Driver-IO/dashboard/dashboard.ino

```diff
@@ -0,0 +1,148 @@
+  pinMode(GEAR_SELECTOR_PIN, INPUT_PULLUP); // Uses built-in pullup resistor.
+
+  pinMode(IMD_INDICATOR_PIN, OUTPUT);
+  pinMode(BMS_INDICATOR_PIN, OUTPUT);
+  pinMode(SPEAKER_PIN, OUTPUT);
+  pinMode(BUTTON_LED_PIN, OUTPUT);
+
+  while (CAN_OK != CAN.begin(CAN_500KBPS, MCP_8MHz)) { //specify 8MHz crystal
+    Serial.println("CAN BUS init Failed");
+    delay(250);
+  }
+  Serial.println("CAN BUS Shield Init OK!");
+}
+
+void imdDetection();
+
+void bmsDetection();
+
+bool forwardReverse();
+
+void onOff();
+
+void changeDireciton();
+
+void loop() {
+  imdDetection();
+  bmsDetection();
+  onOff();
+  changeDirection();
+}
+
+/**
+   Turns IMD indicator on or off depending on if IMD_INPUT_PIN detects a voltage.
+*/
+void imdDetection() {
+  if (digitalRead(IMD_INPUT_PIN)) {
+    digitalWrite(IMD_INDICATOR_PIN, HIGH);
+  } else {
+    digitalWrite(IMD_INDICATOR_PIN, LOW);
+  }
+}
+
+/**
+   Turns BMS indicator on or off depending on if BMS_INPUT_PIN detects a voltage.
+*/
+void bmsDetection() {
+  if (digitalRead(BMS_INPUT_PIN)) {
+    digitalWrite(BMS_INDICATOR_PIN, HIGH);
+  } else {
+    digitalWrite(BMS_INDICATOR_PIN, LOW);
+  }
+}
+
+/**
+   Use the shutdown circuit detection, on/off button, and gear selection pin to determine the car's state (on or off)
+   and direction. Car will only be ready to drive if shutdown circuit is detected. Speaker is turned on for 3 seconds
+   or until car is turned off (whichever comes first).
+*/
+void onOff() {
+  shutdownVoltage = digitalRead(SHUTDOWN_DETECTION_PIN); // Detects if shutdown circuit is active (5v divided from 12v).
+
+  // If shutdown circuit is active, turn on the LED. Otherwise, turn LED and speaker off, and set state to off
+  if (shutdownVoltage) {
+    digitalWrite(BUTTON_LED_PIN, HIGH);
+  } else {
+    digitalWrite(BUTTON_LED_PIN, LOW);
+    state = 0;
+    speakerCooldown = millis() + 1;
+  }
+
+  // Turns speaker off once cooldown is passed or state is false.
+  if (speakerCooldown < millis() || !state) {
+    analogWrite(SPEAKER_PIN, 0);
+  }
+
+  /* If the shutdown circuit is active, button cooldown has passed, and button is pressed,
+     toggle state, reset button cooldown. If state is true, turn on speaker and tell pedal box
+     to turn on and the drive direction of the vehicle. If car was turned off, tell that to the pedal box.
+  */
+  if (shutdownVoltage && buttonCooldown < millis() && digitalRead(BUTTON_PIN) == LOW) {
+    state = !state;
+    buttonCooldown = millis() + 500; // Set cooldown to 500ms (prevents triggering button multiple times with one press).
+    if (state) { // Car is turned on.
+      analogWrite(SPEAKER_PIN, 128);
+      speakerCooldown = millis() + 3000;
+      CAN.sendMsgBuf(0x01, 0, 1, ON); // Tell pedal box to turn inverter on
+      if (digitalRead(GEAR_SELECTOR_PIN)) { // Send direction to pedal box when car is turned on.
+        CAN.sendMsgBuf(0x02, 0, 1, FORWARD);
+      } else {
+        CAN.sendMsgBuf(0x02, 0, 1, REVERSE);
+      }
```

Comment:
I spy duplicate code. Same ideas from before apply. Good in principle, not necessary in practice.